service/src/main/java/bio/terra/pipelines/app/controller/JobApiUtils.java

```diff
@@ -121,13 +134,97 @@ public static ApiErrorReport buildApiErrorReport(Exception exception) {
     }
   }
 
-  private static String resultUrlFromFlightState(FlightState flightState) {
+  /**
+   * Retrieves the result of an asynchronous job.
+   *
+   * <p>Stairway has no concept of synchronous vs asynchronous flights. However, MC Terra has a
+   * service-level standard result for asynchronous jobs which includes a ApiJobReport and either a
+   * result or error if the job is complete. This is a convenience for callers who would otherwise
+   * need to construct their own AsyncJobResult object.
+   *
+   * <p>Unlike retrieveJobResult, this will not throw for a flight in progress. Instead, it will
+   * return a ApiJobReport without a result or error.
+   */
+  public <T> AsyncJobResult<T> retrieveAsyncJobResult(
+      UUID jobId,
+      String userId,
+      PipelinesEnum pipelineId,
+      Class<T> resultClass,
+      TypeReference<T> typeReference) {
+    try {
+      FlightState flightState = jobService.retrieveJob(jobId, userId, pipelineId);
+      ApiJobReport jobReport = mapFlightStateToApiJobReport(flightState);
+      if (jobReport.getStatus().equals(ApiJobReport.StatusEnum.RUNNING)) {
+        return new AsyncJobResult<T>().jobReport(jobReport);
+      }
+
+      // Job is complete, get the result
+      JobService.JobResultOrException<T> resultOrException =
+          jobService.retrieveJobResult(jobId, resultClass, typeReference);
+      final ApiErrorReport errorReport;
+      if (jobReport.getStatus().equals(ApiJobReport.StatusEnum.FAILED)) {
+        errorReport = buildApiErrorReport(resultOrException.getException());
+      } else {
+        errorReport = null;
+      }
+      return new AsyncJobResult<T>()
+          .jobReport(jobReport)
+          .result(resultOrException.getResult())
+          .errorReport(errorReport);
+    } catch (StairwayException stairwayEx) {
+      throw new InternalStairwayException(stairwayEx);
+    }
+  }
+
+  private String resultUrlFromFlightState(FlightState flightState) {
```

Comment:
so it looks like 2 of these functions are non static and the rest (4?) are static.  Is there any meaningful way we can split them up into two separate files?  If not i would probably leave the static ones as static.  Theres enough of them that its meaningful to have them be "correct" i think.

I wonder if the two non static methods could be moved into JobService?